### PR TITLE
fixes #1646, #3103 - enable cloning for templates, and add read-only options

### PR DESCRIPTION
**Pull request by stbenjam**

### CODE REVIEW

File: app/models/config_template.rb [hunk 1]
```diff
@@ -41,6 +42,12 @@ def to_param
     "#{id}-#{name.parameterize}"
   end
 
+  def clone
+    template = self.dup
```
Comment (dLobatog) [inline app/models/config_template.rb]: Could you use deep_cloneable to clone the relationships with operating systems at least? https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L645 is an example

File: app/views/config_templates/index.html.erb [hunk 2]
```diff
@@ -22,8 +22,8 @@
       <td><%= combination config_template %></td>
       <td><%= config_template.try(:template_kind) %></td>
       <td><%= checked_icon config_template.snippet %></td>
-      <td><%= display_delete_if_authorized hash_for_config_template_path(:id => config_template.to_param).merge(:auth_object => config_template, :authorizer => authorizer, :permission => 'destroy_templates'),
-                                           :confirm => _("Delete %s?") % config_template %> </td>
+      <td><%= action_buttons(display_link_if_authorized(_('Clone'), hash_for_clone_config_template_path(:id => config_template)),
+                             display_delete_if_authorized(hash_for_config_template_path(:id => config_template.to_param).merge(:auth_object => config_template, :authorizer => authorizer, :permission => 'destroy_templates'), :confirm => _("Delete %s?") % config_template)) %> </td>
```
Comment (dLobatog) [inline app/views/config_templates/index.html.erb]: Split this in a couple of lines please, it'd be easier to read without scrolling :smile:

File: test/unit/config_template_test.rb [hunk 3]
```diff
@@ -64,6 +64,20 @@ def test_should_require_a_template_kind
     assert !tmplt.save
   end
 
+  def test_should_be_able_to_clone
```
Comment (dLobatog) [inline test/unit/config_template_test.rb]: Could you make it `def test` to `test "xxxx" do end`? Same for all tests in the file, sorry they weren't your fault :smile:

File: app/controllers/config_templates_controller.rb [hunk 4]
```diff
@@ -75,4 +95,20 @@ def default_template_url template, hostgroup
...
+  def action_permission
+    case params[:action]
+      when 'new', 'create'
+        'create'
+      when 'edit', 'update'
+        'edit'
+      when 'destroy'
+        'destroy'
+      when 'index', 'show', 'clone'
+        'view'
```
Comment (jlsherrill) [inline app/controllers/config_templates_controller.rb]: Shouldn't clone require the 'create' permission (even though i realize its not doing much but redirecting to new)?

**Comment by jlsherrill** [comment]
When i go to clone a template, the org and location list is empty.  I would expect it to have whatever org and location I'm in similar to when i create a new template.

File: app/controllers/config_templates_controller.rb (hunk 4 above)
Comment (stbenjam) [inline app/controllers/config_templates_controller.rb]: Host and Host group both use 'view' for clone.  I assume because the permission is evaluated on the original object, and you only need to view that one. 

Although it is confusing to show a clone link if the user ultimately doesn't have create rights.

**Comment by stbenjam** [comment]
@jlsherrill What should the behavior be? At the moment, I clone the org and locs from the previous template.  I guess I should take only the current taxonomies, if set?

File: app/controllers/config_templates_controller.rb (hunk 4 above)
Comment (jlsherrill) [inline app/controllers/config_templates_controller.rb]: ah, then i'm okay with that if the foreman-ers are okay with it :)

**Comment by jlsherrill** [comment]
@stbenjam i could see an argument for either/both.  

a) Since its a clone you'd expect everything I don't change to be the same (the way you've implemented it).  
b) As a non-admin user, I might not even have permissions to access those other orgs and locations though (I assume an error would be thrown on the save?).  
c) As a user I might expect it to behave the same as new.
d) As a katello user, I would want it to select my current selected org/location as that is more our workflow (user clones to their org to make edits).

For the katello usecase, katello can override the ConfigTemplate#clone method if need be.  After thinking about it, i'd probably lean towards a) However I'm not sure about situation b).

File: app/controllers/config_templates_controller.rb (hunk 4 above)
Comment (domcleal) [inline app/controllers/config_templates_controller.rb]: It's correct, though the link should only be visible if the user has the create_templates permission (due to access_permissions.rb).

File: app/views/config_templates/_form.html.erb [hunk 5]
```diff
@@ -6,15 +6,32 @@
...
       <div class="tab-pane active" id="primary">
-        <%= text_f f, :name %>
+        <%= text_f f, :name, :disabled => !@config_template.editable? %>
 
+        <% if !@config_template.editable? -%>
+          <% allowed_changes = [_("associations")]
+             allowed_changes << _("locations")     if show_location_tab?
+             allowed_changes << _("organizations") if show_organization_tab? %>
+        <%= alert :class => 'alert-warning', :header => '',
+                  :text  => icon_text("warning-sign", (_('Note: This template is read only.  You may only change the %s. Please %s it to customize.') %
```
Comment (domcleal) [inline app/views/config_templates/_form.html.erb]: Use hash syntax when multiple variables are interpolated: http://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers

Two nitpicks, "read only" should be hyphenated, one space after a period for consistency.

File: app/models/config_template.rb [hunk 6]
```diff
@@ -133,8 +141,32 @@ def skip_strip_attrs
...
+    editable || migration?
+  end
+
+  def removable?
+    unless removable || migration?
+      errors.add(:base, _("This template is not removable."))
+    end
+  end
+
+  def migration?
```
Comment (domcleal) [inline app/models/config_template.rb]: Use `Foreman.in_rake?` instead.

File: app/models/config_template.rb [hunk 7]
```diff
@@ -133,8 +141,32 @@ def skip_strip_attrs
     ['template']
   end
 
+  def editable?
+    editable || migration?
+  end
+
+  def removable?
+    unless removable || migration?
+      errors.add(:base, _("This template is not removable."))
```
Comment (domcleal) [inline app/models/config_template.rb]: cannot be deleted?  is protected and cannot be removed?

File: app/models/concerns/taxonomix.rb [hunk 8]
```diff
@@ -159,7 +159,7 @@ def taxonomy_foreign_key_conditions
 
   def used_taxonomy_ids(type)
     return [] if new_record?
-    Host::Base.where(taxonomy_foreign_key_conditions).pluck(type).uniq.compact
+    Host::Base.where(taxonomy_foreign_key_conditions).pluck(type).uniq.compact rescue []
```
Comment (domcleal) [inline app/models/concerns/taxonomix.rb]: Why/what is this rescuing?

File: app/controllers/config_templates_controller.rb [hunk 9]
```diff
@@ -75,4 +95,20 @@ def default_template_url template, hostgroup
   def controller_permission
     'templates'
   end
+
+  def action_permission
+    case params[:action]
+      when 'new', 'create'
+        'create'
```
Comment (domcleal) [inline app/controllers/config_templates_controller.rb]: You can usually avoid duplicating this by doing:

<pre>
def action_permission
  case params[:action]
  when 'clone'
    :view
  else
    super
  end
end
</pre>

File: app/assets/javascripts/config_template.js [hunk 10]
```diff
@@ -104,6 +104,9 @@ function create_editor(item) {
   $editor = ace.edit("editor1");
   $editor.setShowPrintMargin(false);
   $editor.renderer.setShowGutter(false);
+  if (item.is(':disabled')) {
+      $('.ace_text-input').attr('disabled', true)
```
Comment (domcleal) [inline app/assets/javascripts/config_template.js]: 2-space indentation

File: test/unit/config_template_test.rb [hunk 11]
```diff
@@ -64,6 +64,52 @@ def test_should_require_a_template_kind
...
+    assert_equal clone.operatingsystems, tmplt.operatingsystems
+    assert_equal clone.template_kind_id, tmplt.template_kind_id
+    assert_equal clone.template, tmplt.template
+    assert_equal tmplt.editable, false
+    assert_equal clone.editable, true
+  end
+
+  test "should not remove an unremovable template" do
+    tmplt = config_templates(:unremovable)
+    assert !tmplt.destroy
```
Comment (domcleal) [inline test/unit/config_template_test.rb]: There were some helpers added recently, so you can do things like: `refute_with_errors u.destroy, tmplt, :base, /removable/`

File: test/unit/config_template_test.rb [hunk 12]
```diff
@@ -64,6 +64,52 @@ def test_should_require_a_template_kind
...
+    assert_equal clone.operatingsystems, tmplt.operatingsystems
+    assert_equal clone.template_kind_id, tmplt.template_kind_id
+    assert_equal clone.template, tmplt.template
+  end
+
+  test "should not edit an uneditable template" do
+    tmplt = config_templates(:uneditable)
+    tmplt.name = "something else"
+    assert !tmplt.valid?
+    assert tmplt.errors[:base].first.include?('is read-only')
```
Comment (domcleal) [inline test/unit/config_template_test.rb]: ditto `refute_valid tmplt, :base, /is read-only/`

**Comment by domcleal** [comment]
Integration tests are failing, I'd guess you have a SQL query error occurring which is invalidating the txn.  Could you also add a functional test for the clone route?

File: app/models/concerns/taxonomix.rb (hunk 8 above)
Comment (stbenjam) [inline app/models/concerns/taxonomix.rb]: Good question :grin:  

It's assuming that each taxonomic model has a host association (?), and this blows up for config templates.  I tried returning [] if Host::Base.column_names didn't include the keys in taxonomy_foreign_key_conditions, but that caused the Smart Proxy taxonomy tests to explode...

Any ideas what I should do? I'm not even really sure what these used_???_ids methods are supposed to be doing.

Comment (domcleal) [inline app/models/concerns/taxonomix.rb]: I think we'd need to reproduce the issue, maybe @isratrade has some time to look into it?

Comment (stbenjam) [inline app/models/concerns/taxonomix.rb]: It's probably also related to the current test failures, which stem from the loc/org tab view.

**app/views/taxonomies/_loc_org_tabs.html.erb:5:**

<pre>
 {'data-descendants' => obj.children_of_selected_location_ids.to_json,
</pre>

File: app/controllers/config_templates_controller.rb [hunk 13]
```diff
@@ -11,9 +11,19 @@ def index
   end
 
   def new
+    @operatingsystems = []
     @config_template = ConfigTemplate.new
   end
 
+  def clone
+    @clone = @config_template.clone
+    @config_template = @clone
```
Comment (dLobatog) [inline app/controllers/config_templates_controller.rb]: `@clone` seems unnecessary here, `@config_template = @config_template.clone` would work too.

Comment (stbenjam) [inline app/controllers/config_templates_controller.rb]: I took this verbatim from the existing clone methods in other models.  I assumed they did this for a reason, but you're right that should work.

**Comment by GregSutcliffe** [comment]
Overall +1, I think marking our default templates as read-only is fine (we can disable editing, but still perhaps allow complete deletion if the user really doesn't care?)

I'd suggest a boolean flag for distributor templates vs user templates, and then the distributor ones could be shown differently in the index, perhaps a separate tab or graying out of lines? Clicking on 5 templates trying to find one you can edit isn't helpful...

**Comment by stbenjam** [comment]
@GregSutcliffe A tab looks like it won't be too simple to implement, maybe a column? 

A user can also search "editable = True" if we enable the field for search.

**Comment by ohadlevy** [comment]
how about a generic lock option? (make the unlock a separate process?)

**Comment by stbenjam** [comment]
@ohadlevy I hadn't planned to make them user-unlockable, the reasoning is to support default templates that can be added to multiple orgs.  We don't want a user in Org A to be able to edit "Kickstart default RHEL" and affect a user in Org B.  They'll have to clone to customize.  But we want both Org A and Org B to see all the default templates.  Then when a user creates Org C, Katello will assign all the templates that have this 'distributor' flag that @GregSutcliffe mentioned to that org too.

I also didn't plan to change Foreman's way of doing things, because locking the default templates breaks users who _have_ edited them already, and people who use the community-templates plug-in.

Do you still want the ability to lock/unlock them in the  UI by an admin? I think we'd end up having to remove that in Katello, we want to be able to always control the default templates.

File: app/controllers/config_templates_controller.rb [hunk 14]
```diff
@@ -24,6 +34,7 @@ def create
   end
 
   def edit
+    load_vars_for_ajax
```
Comment (ohadlevy) [inline app/controllers/config_templates_controller.rb]: Hmm.. what is the reason why need this ? AFAIR we dont have any ajax in the template editor?

File: bundler.d/assets.rb [hunk 15]
```diff
@@ -3,7 +3,7 @@
   gem 'uglifier', '>= 1.0.3'
   gem 'execjs', '< 2.1.0'
   gem "jquery-rails", "2.0.3"
-  gem 'jquery-ui-rails'
+  gem 'jquery-ui-rails', '< 5.0.0'
```
Comment (ohadlevy) [inline bundler.d/assets.rb:6]: just out of curiosity, why?

File: app/views/config_templates/_form.html.erb [hunk 16]
```diff
@@ -6,15 +6,33 @@
       <li><a href="#template_type" data-toggle="tab"><%= _("Type") %></a></li>
       <li><a href="#template_associations" data-toggle="tab"><%= _("Association") %></a></li>
       <li><a id='history_tab' href="#history" data-toggle="tab"><%= _("History") %></a></li>
+      <% if show_location_tab? %>
```
Comment (ohadlevy) [inline app/views/config_templates/_form.html.erb]: is this related? I dont mind, but it sounds like a separate PR (maybe there are other objects which are missing the tabs)

File: app/views/config_templates/_form.html.erb [hunk 17]
```diff
@@ -6,15 +6,33 @@
...
+      <% end %>
+      <% if show_organization_tab? %>
+        <li><a href="#organizations" data-toggle="tab"><%= _("Organizations") %></a></li>
+      <% end %>
     </ul>
 
     <div class="tab-content">
       <div class="tab-pane active" id="primary">
-        <%= text_f f, :name %>
+        <%= text_f f, :name, :disabled => !@config_template.editable? %>
```
Comment (ohadlevy) [inline app/views/config_templates/_form.html.erb]: does it make sense to add a new attribute to identify the template ? I would like down the road the ability to auto attach a template to OS's based on metadata in the template?

File: app/views/config_templates/_form.html.erb [hunk 18]
```diff
@@ -6,15 +6,33 @@
...
 
+        <% if !@config_template.editable? -%>
+          <% allowed_changes = [_("associations")]
+             allowed_changes << _("locations")     if show_location_tab?
+             allowed_changes << _("organizations") if show_organization_tab? %>
+        <%= alert :class => 'alert-warning', :header => '',
```
Comment (ohadlevy) [inline app/views/config_templates/_form.html.erb:16]: nitpick: indentation, maybe extract to a helper

File: app/views/config_templates/_form.html.erb [hunk 19]
```diff
@@ -89,10 +110,11 @@
             </ul>
             <%= (_("The final entry, Operating System default, can be set by editing the %s page.") % (link_to _("Operating System"), operatingsystems_path)).html_safe %>
           </div>
-          <%= multiple_checkboxes f, :operatingsystems, @config_template, Operatingsystem, :label => _("Applicable<br>Operating Systems").html_safe %>
+          <%= multiple_selects f, :operatingsystems, Operatingsystem, @operatingsystems.try(:map, &:id), :label => _("Applicable<br>Operating Systems").html_safe %>
```
Comment (ohadlevy) [inline app/views/config_templates/_form.html.erb:103]: why?

File: app/controllers/config_templates_controller.rb (hunk 14 above)
Comment (stbenjam) [inline app/controllers/config_templates_controller.rb]: When we load the new form, the associations aren't saved in the database so we have to store them in variables like @operatingsystems, @oraganizations, @locations.  It's not ajax here, but it's how it's named for the similar feature in other models.

File: app/views/config_templates/_form.html.erb (hunk 19 above)
Comment (stbenjam) [inline app/views/config_templates/_form.html.erb:103]: See the explanation above for load_ajax_vars.  This is how the other views have it for cloning.  

multiple_checkboxes tries to find the associations.  When you clone, it's not saved to the DB, so there are no associations.

File: bundler.d/assets.rb (hunk 15 above)
Comment (stbenjam) [inline bundler.d/assets.rb:6]: Because of #1564 ,which is fixed now.  I'll remove this here.

**Comment by GregSutcliffe** [comment]
@stbenjam i'm concerned about minimizing where katello diverges from core, where possible. I don't think we consume templates that differently, so I'd rather design a good workflow now, than end up in another mess later.

I think templates need to be unlockable, but with a big disclaimer about how that means the user won't get updates to that template in future releases. Give power to the user, don't lock them in. At upgrade time, you can then update Templates.where(matching name && locked). Would that be sufficient for Katello too?

re: my earlier point, I'm less concerned about how it looks, whatever works for you, so long as we can visually tell on the index whats editable and what isn't.

**Comment by stbenjam** [comment]
Ok just pushed the latest iteration, taking @ohadlevy's locked suggestion.  I like the usage workflow better.

@GregSutcliffe , I added an icon to indicate locked -- what do you think? We need different icons but just an example.  Maybe one for unlocked too?

By default, only an admin can lock/unlock templates.  I like this because then a company can provide all the organizations a set of "standard" templates for them to use, not just the ones provided by us.

File: app/views/config_templates/index.html.erb [hunk 20]
```diff
@@ -18,13 +18,16 @@
   </tr>
   <% for config_template in @config_templates %>
     <tr>
-      <td><%= link_to_if_authorized h(config_template), hash_for_edit_config_template_path(:id => config_template.to_param).merge(:auth_object => config_template, :authorizer => authorizer, :permission => 'edit_templates') %></td>
+      <td>  <%= image_tag("locked.png") if config_template.locked? %>
```
Comment (ohadlevy) [inline app/views/config_templates/index.html.erb]: use the lock icon we already have --> http://getbootstrap.com/components/

**Comment by GregSutcliffe** [comment]
@stbenjam Looks good, will have to do a proper code reviwe soonish. I agree with your reasons for admin only, but there's no reason not to have a permission admins can grant if they wish - just don't enable it on any roles by default ;)

No need for an unlocked icon, that will clutter and make it hard to differentiate at speed.

**Comment by stbenjam** [comment]
@GregSutcliffe @ohadlevy @eLobato Cool, thanks for all the feedback so far.  I think this is better than the first iteration.

Demo of what it looks like:
locked2

**Comment by stbenjam** [comment]
Last change for now.  Templates which are seeded (by a plug-in for example) may set the 'default' field and lock their templates.  These may not be unlocked, and we're able to let the user know why because the seeder may opt to say who the vendor is for the templates.

locked3

File: app/models/config_template.rb [hunk 21]
```diff
@@ -3,10 +3,14 @@ class ConfigTemplate < ActiveRecord::Base
...
   audited :allow_mass_assignment => true
   self.auditing_enabled = !Foreman.in_rake?('db:migrate')
-  attr_accessible :name, :template, :template_kind, :template_kind_id, :snippet, :template_combinations_attributes, :operatingsystems, :operatingsystem_ids, :audit_comment
+  attr_accessible :name, :template, :template_kind, :template_kind_id, :snippet, :template_combinations_attributes,
+                  :operatingsystems, :operatingsystem_ids, :audit_comment, :location_ids, :organization_ids, :locked,
+                  :vendor, :default
   validates :name, :presence => true, :uniqueness => true
   validates :name, :template, :presence => true
   validates :template_kind_id, :presence => true, :unless => Proc.new {|t| t.snippet }
+  validate :changes_when_template_is_locked
```
Comment (dLobatog) [inline app/models/config_template.rb]:
``` ruby
validate :changes, :if => lambda { |template| template.locked? || template.locked_changed? }
```

And remove the early return on `changes_when_template_is_locked`.

File: app/models/config_template.rb [hunk 22]
```diff
@@ -133,8 +143,36 @@ def skip_strip_attrs
...
 
+  def locked?
+    locked && !Foreman.in_rake?
+  end
+
   private
 
+  def check_if_template_is_locked
+    if locked?
+      errors.add(:base, _("This template is locked and may not be removed."))
```
Comment (dLobatog) [inline app/models/config_template.rb]: I think an inline `if locked?` would fit better here since it's just a one line condition.

File: app/controllers/config_templates_controller.rb [hunk 23]
```diff
@@ -14,6 +14,32 @@ def new
...
+  def clone
+    @config_template = @config_template.clone
+    load_vars_for_form
+    flash[:warning] = _("The marked fields will need reviewing")
+    @config_template.valid?
+    render :action => :new
+  end
+
+  def lock
+    @config_template.locked = true
```
Comment (dLobatog) [inline app/controllers/config_templates_controller.rb]: Since it's nearly the same code, I'd suggest:

``` ruby
def lock
  locker('locked')
end

def unlock
  locker('unlocked')
end

private 

def locker(key)
  @config_template.locked = !@config_template.locked
  if @config_template.save
    process_success :success_msg => _("Template #{key}."), :success_redirect => :back
  else
    process_error
  end
end
```

Comment (dLobatog) [inline app/controllers/config_templates_controller.rb]: I can't think of a proper name for `locker`, `lock_helper` perhaps? Tests pass for the above method extraction.

Comment (stbenjam) [inline app/controllers/config_templates_controller.rb]: Locker sounds good to me. That reminds me, I need to add controller tests now.

**Comment by dLobatog** [comment]
Merged as https://github.com/theforeman/foreman/commit/4a28771dbd07fa87161ef09e5b0ebca2840e18e3 , thank you @stbenjam !
